internal/provisioner/helm_utils.go

```diff
@@ -283,16 +289,66 @@ func (d *helmDeployment) Version() (*model.HelmUtilityVersion, error) {
 	return nil, errors.Errorf("unable to get version for chart %s", d.chartDeploymentName)
 }
 
-func applyGitlabTokenIfPresent(original string) string {
-	if os.Getenv(model.GitlabOAuthTokenKey) == "" {
-		return original
+type gitlabValuesFileResponse struct {
+	Content string `json:"content"`
+}
+
+const temporaryValuesFile string = "/tmp/helm-values.yaml"
+
+func fetchFromGitlabIfNecessary(path string) (string, error) {
+	gitlabKey := os.Getenv(model.GitlabOAuthTokenKey)
+	if gitlabKey == "" {
+		return path, nil
+	}
+
+	valPathURL, err := url.Parse(path)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to parse Helm values file path or URL")
+	}
+
+	// silently allow other public non-Gitlab URLs
+	if !strings.HasPrefix(valPathURL.Host, "gitlab") {
+		return path, nil
+	}
+
+	// if Gitlab, fetch the file using the API
+	path = fmt.Sprintf("%s&private_token=%s", path, gitlabKey)
+
+	resp, err := http.Get(path)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to request the values file from Gitlab")
+	}
+	if resp.StatusCode >= 400 {
+		return "", errors.Errorf("request to Gitlab failed with status: %s", resp.Status)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to read body from Gitlab response")
+	}
+
+	valuesFileBytes := new(gitlabValuesFileResponse)
+	err = json.Unmarshal(body, valuesFileBytes)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to unmarshal JSON in Gitlab response")
+	}
+
+	if _, err = os.Stat(temporaryValuesFile); err == nil { // file exists
+		err = os.Remove(temporaryValuesFile)
+		if err != nil {
+			return "", errors.Wrap(err, "failed to overwrite an existing Helm values file to store the new one")
+		}
 	}
-	// gitlab token is set, so apply it to GitLab values path URLs
-	valPathURL, err := url.Parse(original)
-	if err == nil && strings.HasPrefix(valPathURL.Host, "gitlab") {
-		original = fmt.Sprintf("%s&private_token=$%s",
-			original,
-			model.GitlabOAuthTokenKey)
+
+	content, err := base64.StdEncoding.DecodeString(valuesFileBytes.Content)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to decode base64-encoded YAML file")
 	}
-	return original
+
+	err = ioutil.WriteFile(temporaryValuesFile, []byte(content), 0777)
```

Comment:
nit: the file permission could be something like `0600` instead right? Considering we are downloading something from the internet in certain cases it may not hurt to limit that a bit.

Reply:
Since this runs in a single user context in a container it really shouldn't matter but I'm happy to change it to something more restrictive